File: packages-js/kanban-core/src/rules.ts
/** Bewegungsregeln: Spaltenübergänge, gesperrte Spalten, feste Reihenfolge, WIP-Limits. */
import { ALLOWED, deny, type Decision } from './errors'
import { findColumn, type Board, type Card } from './model'

export interface WipState {
  column_id: string
  count: number
  limit: number | null
  full: boolean
  over: boolean
}

/** Kartenzahl einer Spalte, ohne die gerade bewegte Karte. */
export function columnLoad(board: Board, columnId: string, excludeCardId: string | null = null): number {
  return board.cards.filter((card) => card.column_id === columnId && card.id !== excludeCardId).length
}

export function wipStates(board: Board): WipState[] {
  return board.columns.map((column) => {
    const count = columnLoad(board, column.id)
    const limit = column.wip_limit
    return { column_id: column.id, count, limit, full: limit !== null && count >= limit, over: limit !== null && count > limit }
  })
}

/** Passt eine weitere Karte in die Spalte? Im Warnmodus ja, mit Warnung. */
export function checkCapacity(board: Board, columnId: string, excludeCardId: string | null = null): Decision {
  const column = findColumn(board, columnId)
  if (!column) return deny('UNKNOWN_COLUMN', `Unbekannte Spalte '${columnId}'`)
  if (column.wip_limit === null) return ALLOWED
  if (columnLoad(board, columnId, excludeCardId) < column.wip_limit) return ALLOWED
  if (board.wip_mode === 'warn') return { allowed: true, code: 'OK', message: '', warnings: ['WIP_LIMIT_REACHED'] }
  return deny('WIP_LIMIT_REACHED', `WIP-Limit der Spalte „${column.label}“ (${column.wip_limit}) ist erreicht`)
}

function pairAllowed(board: Board, source: string, target: string): boolean {
  if (board.transitions.mode === 'free') return true
  return board.transitions.allowed.some(([from, to]) => from === source && to === target)
}

/** Spaltenwechsel ohne Kapazität; `source === target` ist ein Umsortieren. */
export function checkTransition(board: Board, source: string, target: string): Decision {
  const policy = board.transitions
  if (!findColumn(board, target)) return deny('UNKNOWN_COLUMN', `Unbekannte Spalte '${target}'`)
  if (source === target) {
    if (policy.fixed_order_columns.includes(source)) return deny('ORDER_FIXED', 'Die Reihenfolge dieser Spalte ist fest')
    return ALLOWED
  }
  if (policy.locked_columns.includes(source)) {
    return deny('COLUMN_LOCKED', 'Karten dieser Spalte können nicht verschoben werden')
  }
  if (!pairAllowed(board, source, target)) {
    return deny('TRANSITION_NOT_ALLOWED', `Übergang von '${source}' nach '${target}' ist nicht erlaubt`)
  }
  return ALLOWED
}

/** Vollständige Prüfung für das Verschieben einer Karte (Übergang, dann WIP). */
export function checkMove(board: Board, card: Card, target: string): Decision {
  const decision = checkTransition(board, card.column_id, target)
  if (!decision.allowed || card.column_id === target) return decision
  return checkCapacity(board, target, card.id)
}

/** Spalten, in die die Karte derzeit verschoben werden darf (Tastatur, Menü). */
export function movableTargets(board: Board, card: Card): string[] {
  return board.columns
    .filter((column) => column.id !== card.column_id && checkMove(board, card, column.id).allowed)
    .map((column) => column.id)
}

File: packages/auditcore_kanban/src/auditcore_kanban/rules.py
"""Movement rules: column transitions, locked columns, fixed order and WIP limits."""

from __future__ import annotations

from dataclasses import dataclass

from .errors import ALLOWED, Decision, deny
from .model import Board, Card


@dataclass(frozen=True)
class WipState:
    column_id: str
    count: int
    limit: int | None

    @property
    def full(self) -> bool:
        return self.limit is not None and self.count >= self.limit

    @property
    def over(self) -> bool:
        return self.limit is not None and self.count > self.limit


def column_load(board: Board, column_id: str, exclude_card_id: str | None = None) -> int:
    """Number of cards in a column, not counting the card being moved."""
    return sum(1 for c in board.cards if c.column_id == column_id and c.id != exclude_card_id)


def wip_states(board: Board) -> tuple[WipState, ...]:
    """Current load against the limit for every column."""
    return tuple(
        WipState(col.id, column_load(board, col.id), col.wip_limit) for col in board.columns
    )


def check_capacity(board: Board, column_id: str, exclude_card_id: str | None = None) -> Decision:
    """Can one more card enter ``column_id``? Warn mode allows it with a warning."""
    column = board.column(column_id)
    if column is None:
        return deny("UNKNOWN_COLUMN", f"Unbekannte Spalte '{column_id}'")
    if column.wip_limit is None:
        return ALLOWED
    if column_load(board, column_id, exclude_card_id) < column.wip_limit:
        return ALLOWED
    if board.wip_mode == "warn":
        return Decision(True, warnings=("WIP_LIMIT_REACHED",))
    return deny(
        "WIP_LIMIT_REACHED",
        f"WIP-Limit der Spalte „{column.label}“ ({column.wip_limit}) ist erreicht",
    )


def check_transition(board: Board, source: str, target: str) -> Decision:
    """Column change rule without capacity (``source == target`` is a reorder)."""
    policy = board.transitions
    if board.column(target) is None:
        return deny("UNKNOWN_COLUMN", f"Unbekannte Spalte '{target}'")
    if source == target:
        if source in policy.fixed_order_columns:
            return deny("ORDER_FIXED", "Die Reihenfolge dieser Spalte ist fest")
        return ALLOWED
    if source in policy.locked_columns:
        return deny("COLUMN_LOCKED", "Karten dieser Spalte können nicht verschoben werden")
    if policy.allowed is not None and (source, target) not in policy.allowed:
        message = f"Übergang von '{source}' nach '{target}' ist nicht erlaubt"
        return deny("TRANSITION_NOT_ALLOWED", message)
    return ALLOWED


def check_move(board: Board, card: Card, target: str) -> Decision:
    """Full rule set for moving ``card`` into ``target`` (transition, then WIP)."""
    decision = check_transition(board, card.column_id, target)
    if not decision.allowed or card.column_id == target:
        return decision
    return check_capacity(board, target, exclude_card_id=card.id)


def movable_targets(board: Board, card: Card) -> tuple[str, ...]:
    """Columns the card may currently be moved into (keyboard/menu helpers)."""
    return tuple(
        col.id
        for col in board.columns
        if col.id != card.column_id and check_move(board, card, col.id).allowed
    )
